src/Data/Blob.hs

```diff
@@ -101,62 +99,33 @@ noLanguageForBlob blobPath = throwError (SomeException (NoLanguageForBlob blobPa
 
 -- | Represents a blobs suitable for diffing which can be either a blob to
 -- delete, a blob to insert, or a pair of blobs to diff.
-newtype BlobPair = BlobPair { getBlobPair :: These Blob Blob }
-  deriving (Eq, Show)
+type BlobPair = Edit Blob Blob
 
 instance FromJSON BlobPair where
-  parseJSON = withObject "BlobPair" $ \o -> do
-    before <- o .:? "before"
-    after <- o .:? "after"
-    case (before, after) of
-      (Just b, Just a)  -> pure $ Diffing b a
-      (Just b, Nothing) -> pure $ Deleting b
-      (Nothing, Just a) -> pure $ Inserting a
-      _                 -> Prelude.fail "Expected object with 'before' and/or 'after' keys only"
-
-pattern Diffing :: Blob -> Blob -> BlobPair
-pattern Diffing a b = BlobPair (These a b)
-
-pattern Inserting :: Blob -> BlobPair
-pattern Inserting a = BlobPair (That a)
-
-pattern Deleting :: Blob -> BlobPair
-pattern Deleting b = BlobPair (This b)
-
-{-# COMPLETE Diffing, Inserting, Deleting #-}
+  parseJSON = withObject "BlobPair" $ \o ->
+    fromMaybes <$> (o .:? "before") <*> (o .:? "after")
+    >>= maybeM (Prelude.fail "Expected object with 'before' and/or 'after' keys only")
 
 maybeBlobPair :: MonadFail m => Maybe Blob -> Maybe Blob -> m BlobPair
-maybeBlobPair a b = case (a, b) of
-  (Just a, Nothing) -> pure (Deleting a)
-  (Nothing, Just b) -> pure (Inserting b)
-  (Just a, Just b)  -> pure (Diffing a b)
-  _                 -> Prologue.fail "expected file pair with content on at least one side"
+maybeBlobPair a b = maybeM (Prologue.fail "expected file pair with content on at least one side") (fromMaybes a b)
 
 languageForBlobPair :: BlobPair -> Language
-languageForBlobPair (Deleting b)  = blobLanguage b
-languageForBlobPair (Inserting b) = blobLanguage b
-languageForBlobPair (Diffing a b)
-  | blobLanguage a == Unknown || blobLanguage b == Unknown
-    = Unknown
-  | otherwise
-    = blobLanguage b
+languageForBlobPair = mergeEdit combine . bimap blobLanguage blobLanguage where
+  combine a b
+    | a == Unknown || b == Unknown = Unknown
+    | otherwise                    = b
```

Comment:
Most of the time we shouldn’t even pattern match on edits, we should use eliminators, traversals, and folds. This avoids us accidentally getting the left/right sides swapped, which is a real risk when using the same type on either side.